lib/smart_proxy_reports/spooled_http_client.rb

```diff
@@ -52,37 +52,44 @@ def stop
       @worker.join
     end
 
+    def connect(http, factory, type, endpoint)
+      processed = 0
+      Dir.glob(spool_path("todo", type + "-*")) do |filename|
+        basename = File.basename(filename)
+        logger.debug "Sending report #{basename}"
+        begin
+          post = factory.create_post(endpoint, File.read(filename))
+          response = http.request(post)
+          logger.info "Report #{basename} sent with HTTP response #{response.code}"
+          logger.debug { "Response body: #{response.body}" }
+          if response.code.start_with?("2")
+            if Proxy::Reports::Plugin.settings.keep_reports
+              spool_move("todo", "done", basename)
+            else
+              FileUtils.rm_f spool_path("todo", basename)
+            end
+          else
+            logger.debug { "Moving failed report to 'fail' spool directory" }
+            spool_move("todo", "done", basename)
+          end
+          processed += 1
+        rescue StandardError => e
+          logger.warn "Unable to send #{basename}, will try on next request: #{e}", e
+          raise if ENV["RACK_ENV"] == "test"
+        end
+      end
+      processed
+    end
+
     def process
       processed = 0
       client = ::Proxy::HttpRequest::ForemanRequest.new
       factory = client.request_factory
       # send all files via a single persistent HTTP connection
       logger.debug "Opening HTTP connection to Foreman"
       client.http.start do |http|
-        Dir.glob(spool_path("todo", "*")) do |filename|
-          basename = File.basename(filename)
-          logger.debug "Sending report #{basename}"
-          begin
-            post = factory.create_post("/api/v2/host_reports", File.read(filename))
-            response = http.request(post)
-            logger.info "Report #{basename} sent with HTTP response #{response.code}"
-            logger.debug { "Response body: #{response.body}" }
-            if response.code.start_with?("2")
-              if Proxy::Reports::Plugin.settings.keep_reports
-                spool_move("todo", "done", basename)
-              else
-                FileUtils.rm_f spool_path("todo", basename)
-              end
-            else
-              logger.debug { "Moving failed report to 'fail' spool directory" }
-              spool_move("todo", "done", basename)
-            end
-            processed += 1
-          rescue StandardError => e
-            logger.warn "Unable to send #{basename}, will try on next request: #{e}", e
-            raise if ENV["RACK_ENV"] == "test"
-          end
-        end
+        processed += connect(http, factory, "facts", "/api/v2/hosts/facts")
```

Comment:
I think this is not actually correct to add a number of processed facts to the number of processed reports. It's better to treat them independently.

UPD: Why we send all the gathered facts from different reports at first and only then send all the reports? Shouldn't they supposed to be treated as one entity (a report + its facts) since we receive them both as part of the same payload?

Reply:
Agreed, this is just a cosmetic logging issue, still a valid one. Also please rename method `connect` to `send` as http client block is the one that opens up the connection.

> Why we send all the gathered facts from different reports

We could do that, but it seems easier to just send all facts first and then reports. Why you think this is an issue? The only problem I see is when we would send reports first and then facts because non-existing hosts would not be created.